doc/src/manual/multi-threading.md

```diff
@@ -213,3 +213,33 @@ therefore a blocking call like other Julia APIs.
 It is very important that the called function does not call back into Julia, as it will segfault.
 
 `@threadcall` may be removed/changed in future versions of Julia.
+
+## Caveats
+
+At this time, most operations in the Julia runtime and standard libraries
+are thread safe.
+However, in some areas work on stabilizing thread support is ongoing.
+Multi-threaded programming has many inherent difficulties, and if a program
+using threads exhibits unusual or undesirable behavior (e.g. crashes or
+mysterious results), thread interactions should typically be suspected first.
+
+There are a few specific limitations and warnings to be aware of when using
+threads in Julia:
+
+  * Base collection types require manual locking if used simultaneously by
+    multiple threads (common examples include `push!` on arrays, or inserting
+    items into a `Dict`).
+  * After a task starts running on a certain thread (e.g. via `@spawn`), it
+    will always be restarted on the same thread after blocking. In the future
+    this limitation will be removed, and tasks will migrate between threads.
+  * `@threads` currently uses a static schedule, using all threads and assigning
+    equal iteration counts to each. In the future the default schedule is likely
+    to change to be dynamic.
+  * The schedule used by `@spawn` is nondeterministic and should not be relied on.
+  * Compute-bound, non-memory-allocating tasks can prevent garbage collection from
+    running in other threads that are allocating memory. In these cases it may
+    be necessary to insert a manual call to `GC.safepoint()` to allow GC to run.
+    This limitation will be removed in the future.
+  * Avoid using finalizers in conjunction with threads, particularly if they
```

Comment:
This item feels like it sticks out in the list as being unclear and unreasonable to state. I think we might need to be more explicit here about that finalizers get run simultaneously with code (kind of always have). I can work on some wording and sample code if that helps.

Reply:
Yes, that would be good, thanks.

Reply:
```suggestion
  * Be aware that some library's finalizers may sometimes break badly if
    threads are enabled (though likely are already broken). This may require
    some transitional work across the ecosystem before threading can be widely
    adopted with confidence. Further details on this are below.
```

```md
## Safe use of Finalizers

Because finalizers can interrupt any code, they must be very careful in how
they interact with any global state. Unfortunately, the main reason that
finalizers are used is to update global state (a pure function is generally
rather pointless as a finalizer). This leads us to a bit of a tricky conundrum.
But one which has a few approaches to dealing with:

1. When singled-threaded, code could call access the internal
`jl_gc_enable_finalizers` function to prevent finalizers from being scheduled
inside a critical region. Internally, this is used inside some functions (such
as our C locks) to prevent recursion when doing certain operations (incremental
package loading, codegen, etc.). Observe that this combination (of an
inefficient spinlock and setting this flag) can be used to make finalizers safe
to work.

2. A second strategy, employed by Base in a couple places, is to explicitly try
to acquire the lock non-recursively. For example, `WeakKeyDict` and
`Distributed.finalize_ref` take this approach (though both only handle the
single-threading case correctly). This example below is updated to work in the
multi-threaded case:

        function finalize_ref(r::AbstractRemoteRef)
            if r.where > 0 # Check if the finalizer is already run
                if islocked(client_refs) || !trylock(client_refs)
                    # delay finalizer for later if we aren't free to acquire the lock
                    finalizer(finalize_ref, r)
                    return nothing
                end
                try # `lock` should always be followed by `try` in correct code
                    if r.where > 0 # Must check again here
                        # Do actual cleanup here
                        r.where = 0
                    end
                finally
                    unlock(client_refs)
                end
            end
            nothing
        end

3. A related third strategy is to use a yield-free queue. We don't currently
have a lock-free queue implemented in base, but
`Base.InvasiveLinkedListSynchronized{T}` is suitable. This can frequently be a
good strategy to use for code with event loops. For example, this strategy is
employed by `Gtk.jl` to manage lifetime ref-counting. In this approach, we
don't do any explicit work inside the `finalizer`, and merely add it to a queue
to be dealt with at a safer time. Actually, `schedule` already uses this, so
defining the finalizer as `x -> @spawn do_cleanup(x)` is one example of this
approach. Note however that this doesn't control which thread `do_cleanup` is
running on here, so `do_cleanup` would still need to acquire a lock. That
doesn't need to be true if you implement your own queue, as you can explicitly
only drain that queue from your thread. Additionally, instead of the `@spawn`
convenience macro, you could manually schedule it on a specific thread:

        t = @task do_cleanup(x)
        ccall(:jl_set_task_tid, Cvoid, (Any, Cint), t, 0) # make sticky to thread with tid=1
        schedule(t)
```

Reply:
That's excellent, thanks. I'm not sure we want to mention `jl_set_task_tid`, since then it might be considered "documented" and we will have to keep it.

Reply:
I’m okay with that, though we might want to make it look less like a hack